docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+### Goals
+
+1. Account for disk usage on the nodes.
+
+2. Compatibility with the most important docker storage backends - devicemapper, aufs and overlayfs
+
+3. Provide a roadmap for enabling disk as a schedulable resource in the future.
+
+4. Provide a plugin interface for extending support to non-default filesystems and storage drivers.
+
+### Non Goals
+
+1. Compatibility with all storage backends. The matrix is pretty large already and the priority is to get disk accounting to on most widely deployed platforms.
+
+2. Support for filesystems other than ext4.
+
+### Introduction
+
+Disk accounting in Kubernetes cluster running with docker is complex because of the plethora of ways in which disk gets utilized by a container.
+
+Disk can be consumed for,
+
+1. Container images
+
+2. Container’s writable layer
+
+3. Container’s logs - when written to stdout/stderr and default logging backend in docker is used.
+
+4. Local volumes - hostPath and emptyDir (TODO: Do git volumes consume local disk?)
+
+Kubelet exposes disk usage for the entire node and the container’s writable layer for aufs docker storage driver. This information is made available to end users via the heapster monitoring pipeline.
+
+#### Image layers
+
+Image layers are shared between containers (COW) and so accounting for images is complicated.
+
+Image layers will have to be accounted as system overhead.
+
+As of today, it is not possible to perform feasibility checks on images before being pulled on the node.
+
+#### Writable Layer
+
+Docker creates a writable layer for every container on the host. Depending on the storage driver, the location and the underlying filesystem of this layer will change.
+
+Any files that the container creates or updates (assuming there are no volumes) will be considered as writable layer usage.
+
+The underlying filesystem is whatever the docker storage directory resides on. It is ext4 by default on most distributions.
+
+#### Container’s logs
+
+Docker engine provides a pluggable logging interface. Kubernetes is currently using the default logging mode which is `local file`. In this mode, the docker daemon stores bytes written by containers to their stdout or stderr, to local disk. These log files are contained in a special directory that is managed by the docker daemon. These logs are exposed via `docker logs` interface which is then exposed via kubelet and apiserver APIs. Currently, there is a hard-requirement for persisting these log files on the disk.
+
+#### Local Volumes
+
+Volumes are slightly different from the other use cases local disk. They are pod scoped. Their lifetime is tied to that of a pod. Due to this property accounting of volumes will also be at the pod level.
+
+As of now, the volume types that use local disk directly are ‘HostPath’, ‘EmptyDir’, and ‘GitRepo’. Everything else is a network based volume.
```

Comment:
Tmpfs gets accounted as part of container's memory usage. So I don't think
they fit into the world of disk accounting.

On Fri, Nov 6, 2015 at 11:13 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/disk-accounting.md
> <https://github.com/kubernetes/kubernetes/pull/16889#discussion_r44209183>
> :
>
> > +
> > +Docker creates a writable layer for every container on the host. Depending on the storage driver, the location and the underlying filesystem of this layer will change.
> > +
> > +Any files that the container creates or updates (assuming there are no volumes) will be considered as writable layer usage.
> > +
> > +The underlying filesystem is whatever the docker storage directory resides on. It is ext4 by default on most distributions.
> > +
> > +#### Container’s logs
> > +
> > +Docker engine provides a pluggable logging interface. Kubernetes is currently using the default logging mode which is `local file`. In this mode, the docker daemon stores bytes written by containers to their stdout or stderr, to local disk. These log files are contained in a special directory that is managed by the docker daemon. These logs are exposed via `docker logs` interface which is then exposed via kubelet and apiserver APIs. Currently, there is a hard-requirement for persisting these log files on the disk.
> > +
> > +#### Local Volumes
> > +
> > +Volumes are slightly different from the other use cases local disk. They are pod scoped. Their lifetime is tied to that of a pod. Due to this property accounting of volumes will also be at the pod level.
> > +
> > +As of now, the volume types that use local disk directly are ‘HostPath’, ‘EmptyDir’, and ‘GitRepo’. Everything else is a network based volume.
>
> downward API and secrets do as well, though they're backed by tmpfs -- how
> does tmpfs fit into this scheme?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16889/files#r44209183>.
>